packages/lodestar/src/network/gossip/gossipsub.ts

```diff
@@ -76,23 +76,38 @@ export class LodestarGossipsub extends Gossipsub {
       return false;
     }
 
-    let isValid;
+    let validationResult;
     let transformedObj: GossipObject;
     try {
       const validatorFn = this.getTopicValidator(topic);
       const objSubnet = this.deserializeGossipMessage(topic, message);
       transformedObj = objSubnet.object;
-      isValid = await validatorFn(transformedObj, objSubnet.subnet);
+      validationResult = await validatorFn(transformedObj, objSubnet.subnet);
     } catch (err) {
-      isValid = false;
+      validationResult = ExtendedValidatorResult.reject;
       this.logger.error(`Cannot validate message from ${message.from}, topic ${message.topicIDs}, error: ${err}`);
     }
+    const isValid = this._processTopicValidatorResult(topic, peer, message, validationResult);
```

Comment:
If it's ignore, it will not be propagated to chain?

 I think you should store validation result in transformed object if it's accept|ignore and afterm,  prevent rebroadcast if it's status is "ignore".
That said, not sure how to achieve this without modifying:
https://github.com/ChainSafe/js-libp2p-gossipsub/blob/master/ts/index.ts#L223

Reply:
To summarize:
- accept - emitSelf and send to peers
- ignore - emitSelf only

Reply:
> If it's ignore, it will not be propagated to chain?
no it'll not propagate for reject & ignore, only propagate for accept.

My understanding is we treat ignore & reject the same in lodestar as invalid messages. The only difference is to trigger a penalty or not and it should be taken care by gossip-js in the next version of `_processTopicValidatorResult`.
The other thing to consider is to implement optional part like `a client MAY queue future blocks for processing at the appropriate slot` in p2p spec, we'll consider this later.
So I don't see we have to modify `emitSelf` and `transformed object`, please help clarify @wemeetagain @mpetrunic

Reply:
> The other thing to consider is to implement optional part like a client MAY queue future blocks for processing at the appropriate slot in p2p spec, we'll consider this later.

But we have both attestation and block queue implemented and we kinda need gossip to propagate future blocks and attestation

Reply:
I think if block is too far in the future then it should fail, otherwise it's covered by the spec by `MAXIMUM_GOSSIP_CLOCK_DISPARITY`. Probably what we want is the block in `current slot`, not future block.

Reply:
We wan't to queue at least 10ish blocks in future (gossiped blocks). So that once we reach head that we determined from current peer statuses, we can just use those queued blocks instead of sending and waiting for exchanging 20 status requests and setting new target head. 

Currently without queue of future blocks we can never reach head because once we exchange all status requests with our peers head is already further away.

Reply:
> The other thing to consider is to implement optional part like a client MAY queue future blocks for processing at the appropriate slot in p2p spec, we'll consider this later.

Yeah, I think we can/should approach this in a separate PR, keep this one super straightforward.

There are a few MAYs that we will want to implement:
- a client MAY queue future blocks for processing at the appropriate slot
- If the proposer_index cannot immediately be verified against the expected shuffling, the block MAY be queued for later processing while proposers for the block's branch are calculated
- a client MAY queue blocks for processing once the parent block is retrieved

> My understanding is we treat ignore & reject the same in lodestar as invalid messages. The only difference is to trigger a penalty or not and it should be taken care by gossip-js in the next version of _processTopicValidatorResult.

This is indeed the default behavior of gossipsub. We'd want to extend this PR in some way to emit / queue blocks in those scenarios. I think it will be easier to tackle this in a separate PR, as there seems to be several possible solutions / directions to making this work.
My immediate thought is we can just queue these blocks directly from the validation function, but ideally we would use the same handling code as in the accept case. This may require additional thinking / refactoring of js-libp2p-gossipsub depending on the solution.